src/modules/navigator/rtl.cpp

```diff
@@ -56,6 +57,156 @@ RTL::on_inactive()
+	}
+
+	// check if a safe point is closer than home or landing
+	int closest_index = 0;
+
+	for (int current_seq = 1; current_seq <= num_safe_points; ++current_seq) {
+		mission_safe_point_s mission_safe_point;
+
+		if (dm_read(DM_KEY_SAFE_POINTS, current_seq, &mission_safe_point, sizeof(mission_safe_point_s)) !=
+		    sizeof(mission_safe_point_s)) {
+			PX4_ERR("dm_read failed");
+			continue;
+		}
+
+		// TODO: take altitude into account for distance measurement
+		dlat = mission_safe_point.lat - global_position.lat;
+		dlon = mission_safe_point.lon - global_position.lon;
+		double dist_squared = dlat * dlat + dlon * dlon;
+
+		if (dist_squared < min_dist_squared) {
+			closest_index = current_seq;
+			min_dist_squared = dist_squared;
+			closest_safe_point = mission_safe_point;
+		}
+	}
+
+	if (closest_index > 0) {
+		_destination_type = RTL_DESTINATION_SAFE_POINT;
+
+		// There is a safe point closer than home/mission landing
+		// TODO: handle all possible mission_safe_point.frame cases
+		switch (closest_safe_point.frame) {
+		case 0: // MAV_FRAME_GLOBAL
+			_destination.lat = closest_safe_point.lat;
+			_destination.lon = closest_safe_point.lon;
+			_destination.alt = closest_safe_point.alt;
+			_destination.yaw = home_landing_position.yaw;
+			break;
+
+		case 3: // MAV_FRAME_GLOBAL_RELATIVE_ALT
+			_destination.lat = closest_safe_point.lat;
+			_destination.lon = closest_safe_point.lon;
+			_destination.alt = closest_safe_point.alt + home_landing_position.alt; // alt of safe point is rel to home
+			_destination.yaw = home_landing_position.yaw;
+			break;
+
+		default:
+			if (_navigator->get_mission_start_land_available()) {
+				mavlink_log_critical(_navigator->get_mavlink_log_pub(), "RTL: unsupp. MAV_FRAME. Landing at mission land.");
+
+			} else {
+				mavlink_log_critical(_navigator->get_mavlink_log_pub(), "RTL: unsupp. MAV_FRAME. Landing at home position.");
+			}
+
+			break;
+		}
+	}
+
+}
+
+void
+RTL::find_RTL_destination()
+{
+	// get home position:
+	home_position_s &home_landing_position = *_navigator->get_home_position();
+
+	switch (rtl_type()) {
+	case RTL_HOME: { // always take home position as landing destination
+			_destination.set(home_landing_position);
+			_destination_type = RTL_DESTINATION_HOME;
+			break;
+		}
+
+
+	case RTL_LAND: // take mission landing as landing destination (if available), otherwise home
+	case RTL_MISSION: { // inverse mission
+			if (_navigator->get_mission_start_land_available()) {
+				_destination.lat = _navigator->get_mission_landing_lat();
+				_destination.lon = _navigator->get_mission_landing_lon();
+				_destination.alt = _navigator->get_mission_landing_alt();
+				_destination_type = RTL_DESTINATION_MISSION_LANDING;
+
+			} else {
+				_destination.set(home_landing_position);
+				_destination_type = RTL_DESTINATION_HOME;
+			}
+
+			break;
+		}
+
+
+	case RTL_CLOSEST: { // choose closest possible landing point (consider home, mission landing and safe points)
```

Comment:
You don't need the additional brackets (only if you declare a variable inside the block):
```suggestion
	case RTL_CLOSEST: // choose closest possible landing point (consider home, mission landing and safe points)
```